### ISSUE TITLE
false-positive: Call to static method PHPUnit\Framework\Assert::assertCount() with int<0, max> and array will always evaluate to true

### ISSUE THREAD

**Issue by dontub** [phpstan/phpstan-phpunit#332]
With the method below phpstan since lately reports the false-positive "Call to static method PHPUnit\Framework\Assert::assertCount() with int<0, max> and array will always evaluate to true."

```php 
/**
 * @param array<int|string> $expectedKeys
 * @param array<mixed> $actual
 */
public static function assertArrayHasSameKeys(array $expectedKeys, array $actual): void {
  \PHPUnit\Framework\Assert::assertCount(count($expectedKeys), $actual);
  // ...
}
```

Used versions:
phpstan/phpstan 2.2.14
phpstan/phpstan-phpunit 2.0.18

**Comment by SanderMuller** [comment]
I cannot reproduce this from the snippet, so there is probably something else in the configuration.

What I ran, with the method copied as written into a file analysed at level 8:

- phpstan 2.2.13, 2.2.14 (the version you reported) and 2.2.x-dev
- phpstan-phpunit 2.0.18 (the version you reported, at the tag) and current HEAD
- PHPUnit 9.6 and 11.5, whose `assertCount()` signatures differ
- with and without `bleedingEdge`
- called as `Assert::assertCount()`, `self::`, `static::` and `$this->`, and with `list<int|string>`

No errors in any of them.

The harness does report, so the rule and the extension are both active.
`assertCount(2, $twoItems)` with `@param array{int, int} $twoItems` gives the expected
"will always evaluate to true".

For what it is worth, the mechanism looks right in your case. phpstan-phpunit specifies
`assertCount($expected, $actual)` as `$expected === count($actual)`. Both sides are `int<0, max>`
here, so the comparison is *maybe* and nothing should be reported, which matches what I see.
Something in your setup must be narrowing one of the two.

Could you post your `phpstan.neon`, including the level, whether `bleedingEdge` is on and which
other extensions you load, plus your PHPUnit version? A repository that still shows it on a fresh
`composer install` would settle it fastest.

**Comment by staabm** [comment]
@dontub you could also try to reproduce the issue in a small reduced repository
